github/data_source_github_repository_custom_properties.go

```diff
@@ -0,0 +1,103 @@
+package github
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-github/v66/github"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceGithubRepositoryCustomProperties() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceGithubOrgaRepositoryCustomProperties,
+
+		Schema: map[string]*schema.Schema{
+			"repository": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of the repository which the custom properties should be on.",
+			},
+			"property": {
+				Type:        schema.TypeSet,
```

Comment:
The way I see most people using this resource is to:

1. Read the custom properties of a repo
2. For each specific property do something based on it's value. Eg, do X to the repo if the `owner` property equals Y, do Z to the repo if the `environment` property equals ...

This means that it should be as easy as possible to get the value of a specific property. IMO that would be best achieved by using a [`TypeMap`](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-types#typemap). Unfortunately that is not supported by v2 of the terraform-plugin-sdk ([see thread](https://github.com/integrations/terraform-provider-github/pull/2107#discussion_r1677557505)). 

The choice is then between [``TypeList``](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-types#typelist) or [``TypeSet``](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-types#typeset). Reading the docs the latter sounds more correct to me, as the order of the custom properties does not matter. We've seen issues regarding this in this provider in the past, eg in #1950, that should be avoided by using `TypeSet`.

It's worth noting that there are some other implications of using `TypeSet` though, namely that they are [non-indexable](https://github.com/hashicorp/terraform-website/issues/1495#issuecomment-887841699). [This common pattern](https://stackoverflow.com/a/59437020) will for example not work, and throw the error below:

```HCL
data "github_repository_custom_properties" "test" {
  repository    = "test-custom-properties"
}
locals {
  # Assuming the repo have a property called "string" set already
  test = data.github_repository_custom_properties.test.property[index(data.github_repository_custom_properties.test.property.*.property_name, "string")]
}
```

![image](https://github.com/user-attachments/assets/6dcc1314-f555-433b-92e3-58f8da65102d)

That does not mean that it's impossible to fetch the value of a custom property though. The snippet below does work, and is IMO more readable. But I think I'd be amiss if I didn't mention this. 

```HCL
data "github_repository_custom_properties" "test" {
  repository    = "test-custom-properties"
}
locals {
  test = one([ for prop in data.github_repository_custom_properties.test.property: prop if prop.property_name == "string" ])
}
```

I personally lean towards using a `TypeSet` to avoid possible weird diffs, but I don't feel strongly about it. If anyone have any opinions here I'm all ears 😄

Reply:
@felixlut have you considered also adding a `github_repository_custom_property` data source to access a singular property? It would allow for more direct access to properties at the cost of increased API usage.

Reply:
[I have](https://github.com/integrations/terraform-provider-github/pull/2316#pullrequestreview-2456483783), and nothing stops future PR:s for doing just that. The actual implementation would be to simply filter it after [this call](https://github.com/integrations/terraform-provider-github/blob/15b61e7ed0a568dd8b0cb0b61492d8896e5c20e5/github/data_source_github_repository_custom_properties.go#L56-L59)